testutils/test_server.go

```diff
@@ -0,0 +1,206 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package testutils
+
+import (
+	"fmt"
+	"runtime"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/uber/tchannel-go"
+	"github.com/uber/tchannel-go/testutils/goroutines"
+)
+
+// A TestServer encapsulates a TChannel server, a client factory, and functions
+// to ensure that we're not leaking resources.
+//
+// TODO: Include an optional relay.
+type TestServer struct {
+	testing.TB
+
+	// Don't embed this, since we'll soon want to introduce a relay.
+	server *tchannel.Channel
+}
+
+// NewTestServer constructs a TestServer.
+func NewTestServer(t testing.TB, opts *ChannelOpts) *TestServer {
+	return &TestServer{
+		TB:     t,
+		server: NewServer(t, opts),
+	}
+}
+
+// WithTestServer creates a new TestServer, runs the passed function, and then
+// verifies that no resources were leaked.
+//
+// TODO: run function twice; once with a relay, once without.
+func WithTestServer(t testing.TB, chanOpts *ChannelOpts, verifyOpts *goroutines.VerifyOpts, f func(*TestServer)) {
```

Comment:
Since the goroutine verification doesn't happen till after the test runs, we can avoid adding a new parameter by instead having functions on `TestServer` to set the goroutine verify opts.